File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/json/schema/uiHelper/UiHelperTest.java
package com.fincity.nocode.kirun.engine.json.schema.uiHelper;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import org.junit.jupiter.api.Test;

class UiHelperTest {

    @Test
    void testDefaultConstructor() {
        UiHelper uiHelper = new UiHelper();
        assertNotNull(uiHelper);
        assertNotNull("", uiHelper.getValidationMessage("any"));
        assertNull(uiHelper.getComponentPreferred());
    }

    @Test
    void testParameterizedConstructor() {
        Map<String, String> messages = new HashMap<>();
        messages.put("required", "This field is required");
        UiHelper uiHelper = new UiHelper(messages, "Dropdown");

        assertEquals("This field is required", uiHelper.getRequiredMessage());
        assertEquals("Dropdown", uiHelper.getComponentPreferred());
    }

    @Test
    void testSetAndGetValidationMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setValidationMessage("custom", "Custom validation message");

        assertEquals("Custom validation message", uiHelper.getValidationMessage("custom"));
    }

    @Test
    void testRequiredMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setRequiredMessage("Field is required");

        assertEquals("Field is required", uiHelper.getRequiredMessage());
    }

    @Test
    void testMinLengthMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setMinLengthMessage("Minimum length is 5");

        assertEquals("Minimum length is 5", uiHelper.getMinLengthMessage());
    }

    @Test
    void testMaxLengthMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setMaxLengthMessage("Maximum length is 10");

        assertEquals("Maximum length is 10", uiHelper.getMaxLengthMessage());
    }

    @Test
    void testPatternMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setPatternMessage("Invalid pattern");

        assertEquals("Invalid pattern", uiHelper.getPatternMessage());
    }

    @Test
    void testMinValueMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setMinValueMessage("Value must be at least 1");

        assertEquals("Value must be at least 1", uiHelper.getMinValueMessage());
    }

    @Test
    void testMaxValueMessage() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setMaxValueMessage("Value cannot exceed 100");

        assertEquals("Value cannot exceed 100", uiHelper.getMaxValueMessage());
    }

    @Test
    void testSetAndGetComponentPreferred() {
        UiHelper uiHelper = new UiHelper();
        uiHelper.setComponentPreferred("RadioButton");

        assertEquals("RadioButton", uiHelper.getComponentPreferred());
    }

    @Test
    void testFromMethod() {
        Map<String, Object> inputMap = new HashMap<>();
        Map<String, String> messages = new HashMap<>();
        messages.put("required", "This field is required");
        inputMap.put("validationMessages", messages);
        inputMap.put("componentPreferred", "RadioButton");

        UiHelper uiHelper = UiHelper.from(inputMap);

        assertNotNull(uiHelper);
        assertEquals("This field is required", uiHelper.getRequiredMessage());
        assertEquals("RadioButton", uiHelper.getComponentPreferred());
    }

    @Test
    void testFromMethodWithNull() {
        assertNull(UiHelper.from(null));
    }
}

File: kirun-js/__tests__/engine/json/schema/uiHelper/uiHelperTest.ts
import { Schema } from '../../../../../src/engine/json/schema/Schema';
import { UiHelper } from '../../../../../src/engine/json/schema/uiHelper/UiHelper';

test('setUiHelper() should correctly set the UiHelper instance', () => {
    const schema = new Schema();
    const uiHelper = new UiHelper().setRequiredMessage('This field is required');

    schema.setUiHelper(uiHelper);

    expect(schema.getUiHelper()).toBeInstanceOf(UiHelper);
    expect(schema.getUiHelper()?.getRequiredMessage()).toBe('This field is required');
});

test('getUiHelper() should return undefined if no UiHelper is set', () => {
    const schema = new Schema();
    expect(schema.getUiHelper()).toBeUndefined();
});

test('setUiHelper() should overwrite the existing UiHelper instance', () => {
    const schema = new Schema();
    const uiHelper1 = new UiHelper().setRequiredMessage('First message');
    const uiHelper2 = new UiHelper().setRequiredMessage('Second message');

    schema.setUiHelper(uiHelper1);
    expect(schema.getUiHelper()?.getRequiredMessage()).toBe('First message');

    schema.setUiHelper(uiHelper2);
    expect(schema.getUiHelper()?.getRequiredMessage()).toBe('Second message');
});

test('UiHelper.from() should create an instance if a valid object is provided', () => {
    const obj = { validationMessages: { required: 'Field is required' } };
    const uiHelper = UiHelper.from(obj);

    expect(uiHelper).toBeInstanceOf(UiHelper);
    expect(uiHelper?.getRequiredMessage()).toBe('Field is required');
});

test('UiHelper.from() should return undefined for null object', () => {
    expect(UiHelper.from(null)).toBeUndefined();
});

test('UiHelper.from() should create an instance with an empty object', () => {
    expect(UiHelper.from({})).toBeInstanceOf(UiHelper);
});

test('setValidationMessage() should update validation message', () => {
    const uiHelper = new UiHelper();
    uiHelper.setValidationMessage('required', 'Updated Required Message');

    expect(uiHelper.getValidationMessage('required')).toBe('Updated Required Message');
});

test('getValidationMessage() should return undefined for an unset validation type', () => {
    const uiHelper = new UiHelper();
    expect(uiHelper.getValidationMessage('nonExistentType')).toBeUndefined();
});

test('setValidationMessage() should work correctly with multiple validation types', () => {
    const uiHelper = new UiHelper()
        .setRequiredMessage('Required!')
        .setMinLengthMessage('Minimum 3 characters')
        .setMaxLengthMessage('Maximum 10 characters')
        .setPatternMessage('Invalid pattern!')
        .setMinValueMessage('Value too low')
        .setMaxValueMessage('Value too high')

    expect(uiHelper.getRequiredMessage()).toBe('Required!');
    expect(uiHelper.getMinLengthMessage()).toBe('Minimum 3 characters');
    expect(uiHelper.getMaxLengthMessage()).toBe('Maximum 10 characters');
    expect(uiHelper.getPatternMessage()).toBe('Invalid pattern!');
    expect(uiHelper.getMinValueMessage()).toBe('Value too low');
    expect(uiHelper.getMaxValueMessage()).toBe('Value too high');
});


test('setComponentPreferred() should set and retrieve component Preferred', () => {
    const uiHelper = new UiHelper().setComponentPreferred('RadioButton');

    expect(uiHelper.getComponentPreferred()).toBe('RadioButton');
});

test('UiHelper.from() should initialize componentPreferred if provided', () => {
    const obj = { componentPreferred: 'Dropdown' };
    const uiHelper = UiHelper.from(obj);

    expect(uiHelper?.getComponentPreferred()).toBe('Dropdown');
});
